fix(amqplib): stop importing from `amqplib` directly in compiled types

dotboris opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

I started getting type checking errors when using node auto instrumentation:

```
$ tsc --skipLibCheck false --incremental false
node_modules/@opentelemetry/instrumentation-amqplib/build/src/utils.d.ts:3:28 - error TS2307: Cannot find module 'amqplib' or its corresponding type declarations.

3 import type * as amqp from 'amqplib';
                             ~~~~~~~~~


Found 1 error in node_modules/@opentelemetry/instrumentation-amqplib/build/src/utils.d.ts:3
```

Turns out that the compiled `utils.d.ts` imports `amqplib` directly which breaks. This worked fine before #1320 which moved `@types/amqplib` as a dev dependency. The PR adjusted most places where `amqplib` was imported but it missed one.

## Short description of the changes

- Make `getConsumePatch`, `getConfirmedPublishPatch` and `getPublishPatch` in `AmqplibInstrumentation` private. They used to be protected.
- Reduce the number of vendor'd types in `types.ts`

The generate `amqplib.d.ts` now looks like this: (no longer imports `./utils` which imports `amqplib`)

```ts
import { InstrumentationBase, InstrumentationNodeModuleDefinition } from '@opentelemetry/instrumentation';
import { AmqplibInstrumentationConfig } from './types';
export declare class AmqplibInstrumentation extends InstrumentationBase {
    protected _config: AmqplibInstrumentationConfig;
    constructor(config?: AmqplibInstrumentationConfig);
    setConfig(config?: AmqplibInstrumentationConfig): void;
    protected init(): InstrumentationNodeModuleDefinition<unknown>;
    private patchConnect;
    private unpatchConnect;
    private patchChannelModel;
    private unpatchChannelModel;
    private getConnectPatch;
    private getChannelEmitPatch;
    private getAckAllPatch;
    private getAckPatch;
    private getConsumePatch;
    private getConfirmedPublishPatch;
    private getPublishPatch;
    private createPublishSpan;
    private endConsumerSpan;
    private endAllSpansOnChannel;
    private callConsumeEndHook;
    private checkConsumeTimeoutOnChannel;
}
//# sourceMappingURL=amqplib.d.ts.map
```

Flarna commented:
Failed test is unrelated => https://github.com/lerna/lerna/issues/3536

blumamir commented:
Thanks for reporting this.
The fix LGTM, but I think the real issue is that `utils.d.ts` is exported in the public interface where it shouldn't

`amqlib.d.ts`:
```ts
import { InstrumentationBase, InstrumentationNodeModuleDefinition } from '@opentelemetry/instrumentation';
import { AmqplibInstrumentationConfig, ConsumeMessage, Options, Replies } from './types';
import { InstrumentationConsumeChannel, InstrumentationPublishChannel } from './utils';
export declare class AmqplibInstrumentation extends InstrumentationBase {
    protected _config: AmqplibInstrumentationConfig;
    constructor(config?: AmqplibInstrumentationConfig);
    setConfig(config?: AmqplibInstrumentationConfig): void;
    protected init(): InstrumentationNodeModuleDefinition<unknown>;
    private patchConnect;
    private unpatchConnect;
    private patchChannelModel;
    private unpatchChannelModel;
    private getConnectPatch;
    private getChannelEmitPatch;
    private getAckAllPatch;
    private getAckPatch;
    protected getConsumePatch(moduleVersion: string | undefined, original: Function): (this: InstrumentationConsumeChannel, queue: string, onMessage: (msg: ConsumeMessage | null) => void, options?: Options.Consume | undefined) => Promise<Replies.Consume>;
    protected getConfirmedPublishPatch(moduleVersion: string | undefined, original: Function): (this: InstrumentationConsumeChannel, exchange: string, routingKey: string, content: Buffer, options?: Options.Publish | undefined, callback?: ((err: any, ok: Replies.Empty) => void) | undefined) => boolean;
    protected getPublishPatch(moduleVersion: string | undefined, original: Function): (this: InstrumentationPublishChannel, exchange: string, routingKey: string, content: Buffer, options?: Options.Publish | undefined) => boolean;
    private createPublishSpan;
    private endConsumerSpan;
    private endAllSpansOnChannel;
    private callConsumeEndHook;
    private checkConsumeTimeoutOnChannel;
}
//# sourceMappingURL=amqplib.d.ts.map
```

I think a possible alternative fix will be to make the `getConsumePatch`, `getConfirmedPublishPatch` and `getPublishPatch` functions private (I don't see a good reason for them to be protected anyway). That would remove them from the public API which will also remove the `import { InstrumentationConsumeChannel, InstrumentationPublishChannel } from './utils';` part

This will reduce the number of types we need to copy into the instrumentation code, which hopefully can make it a bit simpler to maintain. WDYT? @dotboris @Flarna

Flarna commented:
moving more APIs to private sounds good. can be also done as followup depending on @dotboris motivation to rework this PR.

blumamir commented:
> can be also done as followup depending on @dotboris motivation to rework this PR.

If I got it right, that would mean we don't need to vendor any new amqplib type in `types.ts` in this PR (`export interface Channel extends EventEmitter {` etc).

dotboris commented:
I'll give making the methods private a shot to see how it goes.

dotboris commented:
> Thanks for updating the PR. Do we still need the changes in `types.ts`? (all the new vendored types)

I have reverted the additional types that I previously vendored. I've also removed a bunch of types that did not need to be there anymore. The rest of the vendored types are required because they're used as part of the instrumentation config.

blumamir reviewed (approved):
LGTM, thanks for the fix 🙏
